File: app/src/main/java/com/example/lucas/haushaltsmanager/entities/Booking/ParentBooking.java
package com.example.lucas.haushaltsmanager.entities.Booking;

import androidx.annotation.NonNull;

import com.example.lucas.haushaltsmanager.entities.Price;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;
import java.util.UUID;

public class ParentBooking implements IBooking {
    private final UUID id;
    private Calendar date;
    private final String title;
    private final List<Booking> children;

    public ParentBooking(
            @NonNull String title
    ) {
        this(UUID.randomUUID(), title, Calendar.getInstance(), new ArrayList<>());
    }

    public ParentBooking(
            @NonNull UUID id,
            @NonNull String title,
            @NonNull Calendar date,
            @NonNull List<Booking> children
    ) {
        this.id = id;
        this.title = title;
        this.date = date;
        this.children = children;
    }

    public UUID getId() {
        return id;
    }

    public Calendar getDate() {
        return date;
    }

    public void setDate(Calendar date) {
        this.date = date;
    }

    public String getTitle() {
        return title;
    }

    @Override
    public boolean equals(Object obj) {
        if (!(obj instanceof ParentBooking)) {
            return false;
        }

        ParentBooking other = (ParentBooking) obj;

        return other.getTitle().equals(getTitle())
                && other.getChildren().equals(getChildren());
    }

    public Price getPrice() {
        return new Price(calcChildrenPrice());
    }

    public void addChild(Booking booking) {
        if (children.contains(booking)) {
            return;
        }

        children.add(booking);
    }

    public List<Booking> getChildren() {
        return children;
    }

    private double calcChildrenPrice() {
        double calcPrice = 0;
        for (Booking child : children) {

            calcPrice += child.getSignedPrice();
        }

        return calcPrice;
    }
}

File: app/src/main/java/com/example/lucas/haushaltsmanager/entities/booking/ParentBooking.kt
package com.example.lucas.haushaltsmanager.entities.booking

import androidx.room.Entity
import androidx.room.Ignore
import androidx.room.PrimaryKey
import com.example.lucas.haushaltsmanager.entities.Price
import java.util.*

@Entity(tableName = "parent_bookings")
class ParentBooking(
    @PrimaryKey var id: UUID,
    private var date: Calendar,
    val title: String,
    @Ignore val children: ArrayList<Booking>
): IBooking {

    init {
        for(child in children) {
            child.parentId = id
        }
    }

    constructor(title: String) : this(
        UUID.randomUUID(),
        Calendar.getInstance(),
        title,
        ArrayList<Booking>()
    )

    fun addChildren(bookings: List<Booking>) {
        for (booking in bookings) {
            addChild(booking)
        }
    }

    fun addChild(booking: Booking) {
        if (children.contains(booking)) {
            return
        }

        booking.parentId = id;
        children.add(booking)
    }

    fun getPrice(): Price {
        var price = 0.0
        for (child in children) {
            price += child.price.price
        }

        return Price(price)
    }

    override fun getDate(): Calendar {
        return date
    }

    fun setDate(date: Calendar) {
        this.date = date
    }

    override fun equals(other: Any?): Boolean {
        if (other !is ParentBooking) {
            return false
        }

        return title == other.title
                && children == other.children
    }

    override fun hashCode(): Int {
        var result = id.hashCode()
        result = 31 * result + date.hashCode()
        result = 31 * result + title.hashCode()
        result = 31 * result + children.hashCode()
        return result
    }
}